File: android/app/src/main/java/com/curiousbeagle/android/myaiagent/net/ModelProvider.kt
package com.curiousbeagle.android.myaiagent.net

import kotlinx.coroutines.flow.Flow

/**
 * Events the agent engine consumes — provider-neutral, mirroring iOS
 * `StreamEvent`, so the engine is testable with a fake and a different
 * backend could plug in behind the same seam.
 */
sealed interface StreamEvent {
    data object TextStarted : StreamEvent
    data class TextDelta(val text: String) : StreamEvent
    data class ToolUseStarted(val id: String, val name: String) : StreamEvent
    data class ToolInputDelta(val id: String, val partialJson: String) : StreamEvent
    data class ToolUseFinished(val id: String) : StreamEvent
    data class Finished(val stopReason: StopReason) : StreamEvent

    sealed interface StopReason {
        data object EndTurn : StopReason
        data object ToolUse : StopReason
        data object MaxTokens : StopReason
        data class Other(val raw: String) : StopReason
    }
}

/**
 * The network boundary seam — the `SensorTransport` analog. `ClaudeProvider`
 * is the production implementation; tests use `FakeModelProvider` with
 * scripted events.
 */
fun interface ModelProvider {
    fun stream(
        system: String,
        tools: List<ToolDefinition>,
        messages: List<WireMessage>,
    ): Flow<StreamEvent>
}

File: iOS/MyAIAgent/MyAIAgent/Networking/ModelProvider.swift
//
//  ModelProvider.swift
//  MyAIAgent
//

import Foundation

/// Events the agent engine consumes — provider-neutral, so the engine is
/// testable with a fake and a different backend could plug in behind the
/// same seam.
nonisolated enum StreamEvent: Equatable, Sendable {
    case textStarted
    case textDelta(String)
    case toolUseStarted(id: String, name: String)
    case toolInputDelta(id: String, partialJSON: String)
    case toolUseFinished(id: String)
    case finished(stopReason: StopReason)

    nonisolated enum StopReason: Equatable, Sendable {
        case endTurn
        case toolUse
        case maxTokens
        case other(String)
    }
}

/// The network boundary seam. `ClaudeProvider` is the production
/// implementation; tests use `FakeModelProvider` with scripted events.
nonisolated protocol ModelProvider: Sendable {
    func stream(
        system: String,
        tools: [ToolDefinition],
        messages: [WireMessage]
    ) -> AsyncThrowingStream<StreamEvent, any Error>
}
